reasoner/resolution/resolver/BoundConcludableResolver.java

```diff
@@ -0,0 +1,415 @@
+        }
+        sendAnswerOrSearchRulesOrFail(fromUpstream, iteration, requestState);
+    }
+
+    private void sendAnswerOrSearchRulesOrFail(Request fromUpstream, int iteration, RequestState.CachingRequestState<?, ConceptMap> requestState) {
+        Optional<AnswerState.Partial.Compound<?, ?>> upstreamAnswer = requestState.nextAnswer().map(AnswerState.Partial::asCompound);
+        if (upstreamAnswer.isPresent()) {
+            /*
+            When we only require 1 answer (eg. when the conjunction is already fully bound), we can short circuit
+            and prevent exploration of further rules.
+
+            One latency optimisation we could do here, is keep track of how many N repeated requests are received,
+            forward them downstream (to parallelise searching for the single answer), and when the first one finds an answer,
+            we respond for all N ahead of time. Then, when the rules actually return an answer to this concludable, we do nothing.
+             */
+            if (requestState.isExploration() && requestState.asExploration().singleAnswerRequired()) {
+                requestState.asExploration().downstreamManager().clearDownstreams();
+                requestState.answerCache().setComplete();
+            }
+            answerToUpstream(upstreamAnswer.get(), fromUpstream, iteration);
+        } else {
+            RequestState.Exploration exploration;
+            if (requestState.isExploration() && !requestState.answerCache().isComplete()) {
+                if ((exploration = requestState.asExploration()).downstreamManager().hasDownstream()) {
+                    requestFromDownstream(exploration.downstreamManager().nextDownstream(), fromUpstream, iteration);
+                } else {
+                    requestState.answerCache().setComplete(); // TODO: The cache should not be set as complete during recursion
+                    failToUpstream(fromUpstream, iteration);
+                }
+            } else {
+                failToUpstream(fromUpstream, iteration);
+            }
+        }
+    }
+
+
+    private void requestFromSubsumer(Request.ToSubsumed fromUpstream, int iteration) {
+        Request toSubsumer = Request.ToSubsumer.create(driver(), fromUpstream.subsumer(),
+                                                       fromUpstream, fromUpstream.partialAnswer());
+        requestFromDownstream(toSubsumer, fromUpstream, iteration);
+    }
+
+    @Override
+    protected void initialiseDownstreamResolvers() {
+        throw TypeDBException.of(ILLEGAL_STATE);
+    }
+
+    public void receiveReiterationQuery(ReiterationQuery.Request request) {
+        driver().execute(actor -> actor.respondToReiterationQuery(request, cache.requiresReiteration()));
+    }
+
+    private void respondToReiterationQuery(ReiterationQuery.Request request, boolean reiterate) {
+        request.onResponse().accept(ReiterationQuery.Response.create(driver(), reiterate));
+    }
```

Comment:
what's the point of the extra method here? we can just respond immediately in `receiveReiterationQuery` no?

Reply:
We can, I thought this would be clearer to show where the response is. If you think not then I'll just remove it.